### PR TITLE
Creates LetFreeExprFn for LetFreeEvaluation

### PR DESCRIPTION
**Pull request by snoble**
I'm starting with the most invasive part of https://github.com/johnynek/bosatsu/pull/488. This creates a Value class that an Evaluator can either treat as a function (`asFn`) or it can pass it a let-free expression for a value so the external function can determine its own strategy for how to produce the output (which may not require full evaluation of the argument Value, eg it may use a cached return value).

### CODE REVIEW

**Comment by snoble** [comment]
I'll add some testing.

**Comment by snoble** [comment]
Note that the main reason this invasiveness exists is because it allows for a lot of reuse of the tooling used by Evaluation.

**Comment by johnynek** [comment]
I'm planning on totally rewriting Evaluation to use #473 internally so that can be a kind of a test for the ability to generate code.

I'm a bit concerned about making Evaluation have a dependency on LetFree because then we have two separate ASTs that this is trying to interact with.

Could we instead plug in a trait of a function to handle the callbacks in a generic way that does not create the dependency back, but allows LetFree to pass in the correct functions?

**Comment by snoble** [comment]
> Could we instead plug in a trait of a function to handle the callbacks in a generic way that does not create the dependency back, but allows LetFree to pass in the correct functions?

I'm finally getting back to this. I think I can figure out how to do that. I'll give it a shot tomorrow. Thanks!

**Comment by snoble** [comment]
I created a trait for FnValue to move the LetFree stuff out of Value. But I ended up adding some boxing because I couldn't figure out how to make FnValue a trait that extends Value and then extend FnValue elsewhere (without unsealing Value).

**Comment by snoble** [comment]
The failures are issues downloading dependencies. I'll see what I can test when the codecov report shows up.

**Comment by snoble** [comment]
I'm going to add test coverage to https://github.com/johnynek/bosatsu/pull/513 instead.

File: core/src/main/scala/org/bykn/bosatsu/Externals.scala [hunk 1]
```diff
@@ -39,6 +39,14 @@ object FfiCall {
     def call(t: rankn.Type): Value = evalFn
   }
 
+  final case class ExprFn(wrapper: (LetFreeEvaluation.LetFreeValue, rankn.Type, LetFreeEvaluation.Cache, LetFreeEvaluation.ToLFV) => Any) extends FfiCall {
+    import LetFreeEvaluation.ExprFnValue
+
+    private[this] def evalExprFn(t: rankn.Type): ExprFnValue = ExprFnValue({ (e1, cache, eval) => Value.ExternalValue(wrapper(e1, t, cache, eval)) })
+
+    def call(t: rankn.Type): Value = new Value.FnValue(evalExprFn(t))
```
Comment (johnynek) [inline core/src/main/scala/org/bykn/bosatsu/Externals.scala:47]: I'm going to merge this, but I'd like to refactor to remove Value depending on LetFree (which I think is a circular dependency).

For instance, I think you just want a `case class FromFn(callFn: rankn.Type => Value) extends FfiCall { def call(t: rankn.Type) = callFn(t) }`

Comment (snoble) [reply core/src/main/scala/org/bykn/bosatsu/Externals.scala:47]: That is much simpler than what I had.

**Pull request merged**